cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -16,10 +16,192 @@
 
 #pragma once
 
+#include <cstdint>
 #include <cudf/strings/string_view.cuh>
+#include <hash/hash_constants.hpp>
+
+#include "cuda_runtime_api.h"
+#include "cudf/types.hpp"
+#include "driver_types.h"
+#include "vector_types.h"
 
 using hash_value_type = uint32_t;
 
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief Core MD5 algorith implementation. Processes a single 512-bit chunk,
+ * updating the hash value so far. Does not zero out the buffer contents.
+ */
+void CUDA_DEVICE_CALLABLE md5_hash_step(md5_intermediate_data* hash_state,
+                                        md5_hash_constants_type const* hash_constants,
+                                        md5_shift_constants_type const* shift_constants)
+{
+  uint32_t A = hash_state->hash_value[0];
+  uint32_t B = hash_state->hash_value[1];
+  uint32_t C = hash_state->hash_value[2];
+  uint32_t D = hash_state->hash_value[3];
+
+  uint32_t* buffer_ints = reinterpret_cast<uint32_t*>(hash_state->buffer);
+
+  for (unsigned int j = 0; j < 64; j++) {
+    uint32_t F, g;
+    switch (j / 16) {
+      case 0:
+        F = (B & C) | ((~B) & D);  // D ^ (B & (C ^ D))
+        g = j;
+        break;
+      case 1:
+        F = (D & B) | ((~D) & C);
+        g = (5 * j + 1) % 16;
+        break;
+      case 2:
+        F = B ^ C ^ D;
+        g = (3 * j + 5) % 16;
+        break;
+      case 3:
+        F = C ^ (B | (~D));
+        g = (7 * j) % 16;
+        break;
+    }
+
+    F = F + A + hash_constants[j] + buffer_ints[g];
+    A = D;
+    D = C;
+    C = B;
+
+    uint32_t shift = shift_constants[((j / 16) * 4) + (j % 4)];
+    B              = B + ((F << shift) | (F >> (32 - shift)));
+  }
+
+  hash_state->hash_value[0] += A;
+  hash_state->hash_value[1] += B;
+  hash_state->hash_value[2] += C;
+  hash_state->hash_value[3] += D;
+
+  hash_state->buffer_length = 0;
+}
+
+template <typename Key>
+struct MD5Hash {
+  using argument_type = Key;
+
+  /**
+   * @brief Core MD5 element processing function
+   */
+  template <typename TKey>
+  void __device__ process(TKey const& key,
+                          uint32_t const len,
```

Comment:
Does `len` need to be a parameter? As far as I can see, it's always `sizeof(T)`, in which case there's no need to make it parameter.

Reply:
I think it can be refactored out, let me try. Originally this was using `size_of` and not including it as a parameter was causing issues.